### PR TITLE
 Migrate CircleCI workflows to GitHub Actions (1/3)

### PR DESCRIPTION
**Pull request by AzfaarQureshi**
### What this PR does

This is PR 1/3 of migrating Cortex's CI from CircleCI to GitHub Actions: 

**[Part 1/3](https://github.com/cortexproject/cortex/pull/3302)** 👈 
* create `test-build-deploy` workflow under `.github/workflows`
* add README
* adding the 3 base jobs
   * `lint`
   * `test`
   * `build`
* these jobs are run async and share no dependencies with each other
* workflow is run when a commit is pushed to `master` or when any PR is opened. However CD jobs will only run on the master branch and will be skipped otherwise
* all jobs in following PRs are dependant on one or more of these jobs finishing

**Part 2/3** 
* adding jobs dependant on `build`
    * `integration`
    * `integration-config-db`
* This PR completes the CI portion of migration. The following PR will finish the remaining CD portion of the migration

**Part 3/3** 
* This PR completes the migration from CircleCI to GitHub Actions
   * `deploy_website` is dependant on `build, test`
   * `deploy` is dependant on `build, test, lint, integration, integration-config-db`
* CD jobs will only run on the master branch and will be otherwise skipped

### Which issue(s) this PR fixes:
Fixes https://github.com/cortexproject/cortex/issues/3274

### Checklist
 - [ ] Tests updated
- [x] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: .github/workflows/test-build-deploy.yml [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    runs-on: ubuntu-latest
+    container:
+      image: quay.io/cortexproject/build-image:update-golang-1.14.9-eb0c8d4d2
+    steps:
+      - name: Checkout Repo
+        uses: actions/checkout@v1
+      - name: Sym Link Expected Path to Workspace
+        run: |
+          mkdir -p /go/src/github.com/cortexproject/cortex
+          ln -s $GITHUB_WORKSPACE/* /go/src/github.com/cortexproject/cortex
```
Comment (AzfaarQureshi) [inline .github/workflows/test-build-deploy.yml:19]: A significant number of commands in the Makefile were hardcoded with an assumed file structure of the CI container. make sure previous commands don’t break a symlink was created instead from the hardcoded “expected” working directory `go/src/github.com/cortexproject/cortex` to the actual working directory `$GITHUB_WORKSPACE`.

File: .github/workflows/test-build-deploy.yml [hunk 2]
```diff
@@ -0,0 +1,117 @@
+name: ci
+on:
+  push:
+    branches: [master]
+    tags:
+      - v[0-9]+.[0-9]+.[0-9]+** # Tag filters not as strict due to different regex system on Github Actions
```
Comment (AzfaarQureshi) [inline .github/workflows/test-build-deploy.yml:6]: As of Oct 2020, GitHub [does not support](https://github.community/t/using-regex-for-filtering/16427/2)regex for tag filtering. The old regex  `/^v[0-9]+(\.[0-9]+){2}(-.+|[^-.]*)$/` was approximated using the available GitHub [filter patterns](https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-syntax-for-github-actions#filter-pattern-cheat-sheet)

File: .github/workflows/test-build-deploy.yml [hunk 3]
```diff
@@ -0,0 +1,117 @@
...
+          mkdir /tmp/images
+          ln -s /tmp/images ./docker-images
+          make BUILD_IN_CONTAINER=false save-images
+      - name: Zip Images
+        run: tar -zcvf images.tar.gz /tmp/images
+      - name: Upload Images Artifact
+        uses: actions/upload-artifact@v2
+        with:
+          name: Docker Images
+          path: ./images.tar.gz
```
Comment (AzfaarQureshi) [inline .github/workflows/test-build-deploy.yml:117]: As of Oct 2020 GitHub Actions does not persist between different jobs in the same workflow. Each job is run on a fresh [virtual environment](https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions/introduction-to-github-actions#runners). As such, we need to upload and download artifacts to share data between jobs. 


Docker Images are zipped before uploading as a workaround. The images contain characters that are illegal in the upload-artifact action.

File: .github/README.md [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  run: |
+    mkdir -p /go/src/github.com/cortexproject/cortex
+    ln -s $GITHUB_WORKSPACE/* /go/src/github.com/cortexproject/cortex
+```
+
+**Sharing Artifacts Between Jobs**
+
+As of October 2020, GitHub Actions do not persist between different jobs in the same workflow. Each job is run on a fresh virtual environment (https://docs.github.com/en/free-pro-team@latest/actions/learn-github-actions/introduction-to-github-actions#runners). As such, we need to upload and download artifacts to share data between jobs.
+
+Artifact	Stored in	Used by	Purpose of storing artifact
```
Comment (pracucci) [inline .github/README.md:70]: I think you wanted to format this as a table, but it's missing the table markdown syntax to render it correctly

File: .github/workflows/test-build-deploy.yml (hunk 2 above)
Comment (pracucci) [reply .github/workflows/test-build-deploy.yml:6]: Looks good enough to me 👍

File: .github/workflows/test-build-deploy.yml [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+      image: quay.io/cortexproject/build-image:update-golang-1.14.9-eb0c8d4d2
+    steps:
+      - name: Checkout Repo
+        uses: actions/checkout@v1
+      - name: Sym Link Expected Path to Workspace
+        run: |
+          mkdir -p /go/src/github.com/cortexproject/cortex
+          ln -s $GITHUB_WORKSPACE/* /go/src/github.com/cortexproject/cortex
+      - name: Generate protobuf files
+        run: "protoc -I /go/src:./vendor/github.com/thanos-io/thanos/pkg:./vendor/github.com/gogo/protobuf:./vendor:./pkg/ruler --gogoslick_out=plugins=grpc,Mgoogle/protobuf/any.proto=github.com/gogo/protobuf/types,:./pkg/ruler ./pkg/ruler/ruler.proto"
```
Comment (pracucci) [inline .github/workflows/test-build-deploy.yml:21]: Why is this required? The `make BUILD_IN_CONTAINER=false check-protos` should already build them. Have you had any issue?

My main concern here is that this is a copy of what we already have in the `Makefile` and we would have to keep it synched whenever we add protobuf definitions.

File: .github/workflows/test-build-deploy.yml [hunk 6]
```diff
@@ -0,0 +1,117 @@
...
+        image: cassandra:3.11
+        env:
+          JVM_OPTS: "-Xms1024M -Xmx1024M"
+        ports:
+        - 9042:9042
+    steps:
+      - name: Checkout Repo
+        uses: actions/checkout@v1
+      - name: Get Dependencies
+        run: go get -v -t -d ./...
```
Comment (pracucci) [inline .github/workflows/test-build-deploy.yml:48]: Could you explain why is this required, please?

File: .github/workflows/test-build-deploy.yml [hunk 7]
```diff
@@ -0,0 +1,117 @@
...
+  push:
+    branches: [master]
+    tags:
+      - v[0-9]+.[0-9]+.[0-9]+** # Tag filters not as strict due to different regex system on Github Actions
+  pull_request:
+jobs:
+  lint:
+    runs-on: ubuntu-latest
+    container:
+      image: quay.io/cortexproject/build-image:update-golang-1.14.9-eb0c8d4d2
```
Comment (pracucci) [inline .github/workflows/test-build-deploy.yml:12]: Do you think we can use something similar to `defaults: &defaults` in the CircleCI yaml, in order to avoid the repetition of `image: quay.io/cortexproject/build-image:update-golang-1.14.9-eb0c8d4d2` and being able to define the build image only once in this file?

File: .github/workflows/test-build-deploy.yml [hunk 8]
```diff
@@ -0,0 +1,117 @@
...
+          name: Ring Protobuf
+          path: pkg/ring/ring.pb.go
+      - uses: actions/upload-artifact@v2
+        with:
+          name: Cortex Protobuf
+          path: pkg/ingester/client/cortex.pb.go
+      - uses: actions/upload-artifact@v2
+        with:
+          name: Rules Protobuf
+          path: pkg/ruler/rules/rules.pb.go
```
Comment (pracucci) [inline .github/workflows/test-build-deploy.yml:105]: Have you figured out if we actually need to upload these artifacts? I know we do it in CircleCI but I have the feeling they're unused (at least I don't remember if we still use them anywhere).

**Review by pracucci — APPROVED** [review]
Very good job! Overall LGTM. I just left few questions here and there, but I can't see any blocker.

File: .github/workflows/test-build-deploy.yml (hunk 8 above)
Comment (AzfaarQureshi) [reply .github/workflows/test-build-deploy.yml:105]: Oh, we included it because this PR [here](https://github.com/cortexproject/cortex/pull/1224) made it seem like there was an explicit decision to save these files. Has the situation changed?

File: .github/README.md (hunk 4 above)
Comment (AzfaarQureshi) [reply .github/README.md:70]: oops 🤦  you are right. Sorry this slipped by!

File: .github/workflows/test-build-deploy.yml (hunk 7 above)
Comment (shovnik) [reply .github/workflows/test-build-deploy.yml:12]: There exists a "defaults" field for GitHub actions, but it automatically applies to all jobs in the workflow. The integration and integration-configs-db jobs that will be added with PR 2/3 of the migration do not use the default container so using a default would require explicitly removing the default container from those jobs by setting them to None.

We would like to reuse setup commands and default containers with something like commands in CircleCI, but this feature is not available yet on GitHub Actions.

Status: https://github.community/t/reusing-sharing-inheriting-steps-between-jobs-declarations/16851

File: .github/workflows/test-build-deploy.yml (hunk 5 above)
Comment (AzfaarQureshi) [reply .github/workflows/test-build-deploy.yml:21]: tl;dr: our mistake mb! It's removed now


We couldn't see that line in the [lint command](https://github.com/cortexproject/cortex/blob/master/Makefile#L141) in the makefile so we added it manually. The line also didn't show up in console output under the Lint section
Screen Shot 2020-10-09 at 3 12 40 PM

so we kept it in. But on a second look I realize I missed the `lint: protos` [line](https://github.com/cortexproject/cortex/blob/master/Makefile#L81)  and if we remove line 20, the `protoc -i` is run in `make lint` and the output shows in the console
Screen Shot 2020-10-09 at 3 16 11 PM

File: .github/workflows/test-build-deploy.yml (hunk 7 above)
Comment (AzfaarQureshi) [reply .github/workflows/test-build-deploy.yml:12]: GHA doesn't support yaml anchors yet because of some security reasons but they are actively working on it: https://github.community/t/reusing-sharing-inheriting-steps-between-jobs-declarations/16851/14

File: .github/workflows/test-build-deploy.yml (hunk 6 above)
Comment (AzfaarQureshi) [reply .github/workflows/test-build-deploy.yml:48]: you are right, this is vestigial code. apologies! next PR will be more thoroughly double checked

File: .github/README.md [hunk 9]
```diff
@@ -0,0 +1,95 @@
...
+
+**Symbolic Link to Expected Workspace**
+
+A significant number of commands in the Makefile are hardcoded with an assumed file structure of the CI container. To ensure paths specified in previous commands don’t break, a symlink was created from the hardcoded “expected” working directory `/go/src/github.com/cortexproject/cortex` to the actual working directory `$GITHUB_WORKSPACE`.
+
+```yaml
+- name: Sym link expected path to github workspace
+  run: |
+    mkdir -p /go/src/github.com/cortexproject/cortex
+    ln -s $GITHUB_WORKSPACE/* /go/src/github.com/cortexproject/cortex
```
Comment (pstibrany) [inline .github/README.md:63]: This looks suspicious... why create many links instead of one:

```
mkdir -p /go/src/github.com/cortexproject/
ln -s "$GITHUB_WORKSPACE" /go/src/github.com/cortexproject/cortex
```

File: .github/README.md [hunk 10]
```diff
@@ -0,0 +1,95 @@
...
+| Docker Images                 | build     | deploy, integration, integrations-config-db | share data between jobs     |
+| Frontend Protobuf             | build     |                                             | long term storage           |
+| Caching Index Client Protobuf | build     |                                             | long term storage           |
+| Ring Protobuf                 | build     |                                             | long term storage           |
+| Rules Protobuf                | build     |                                             | long term storage           |
+
+*Note:* Docker Images are zipped before uploading as a workaround. The images contain characters that are illegal in the upload-artifact action.
+```yaml
+- name: Compressing Images
+        run: tar -zcvf images.tar.gz /tmp/images
```
Comment (pstibrany) [inline .github/README.md:82]: Not sure if compression gives us anything here. We can use `tar` without `-z` flag to make it run faster (but use more disk space).

File: .github/workflows/test-build-deploy.yml [hunk 11]
```diff
@@ -0,0 +1,113 @@
...
+  pull_request:
+jobs:
+  lint:
+    runs-on: ubuntu-latest
+    container:
+      image: quay.io/cortexproject/build-image:update-golang-1.14.9-eb0c8d4d2
+    steps:
+      - name: Checkout Repo
+        uses: actions/checkout@v1
+      - name: Sym Link Expected Path to Workspace
```
Comment (pstibrany) [inline .github/workflows/test-build-deploy.yml:16]: This linking is explained in README, apart from doing review, who reads that? I'd include the explanations from README directly into this file.

File: .github/workflows/test-build-deploy.yml (hunk 7 above)
Comment (pstibrany) [reply .github/workflows/test-build-deploy.yml:12]: Would it be possible to use env variable instead?

File: .github/workflows/test-build-deploy.yml [hunk 12]
```diff
@@ -0,0 +1,113 @@
...
+        run: |
+          mkdir -p /go/src/github.com/cortexproject/cortex
+          ln -s $GITHUB_WORKSPACE/* /go/src/github.com/cortexproject/cortex
+      - name: Lint
+        run: make BUILD_IN_CONTAINER=false lint
```
Comment (pstibrany) [inline .github/workflows/test-build-deploy.yml:21]: Should we set `BUILD_IN_CONTAINER=false` as env variable for entire workflow or at least individual jobs?

File: .github/workflows/test-build-deploy.yml (hunk 8 above)
Comment (pstibrany) [reply .github/workflows/test-build-deploy.yml:101]: Since then we have added these files into Git, so I think #1224 is not relevant anymore.

**Review by pstibrany — APPROVED** [review]
Great job, thank you very much for doing this!

(Nit: In next PRs I would suggest moving the comments from README.md into YAML files, to keep it at one place.)

File: .github/workflows/test-build-deploy.yml (hunk 8 above)
Comment (pstibrany) [reply .github/workflows/test-build-deploy.yml:101]: (In https://github.com/cortexproject/cortex/pull/1399)

File: .github/workflows/test-build-deploy.yml (hunk 7 above)
Comment (pracucci) [reply .github/workflows/test-build-deploy.yml:12]: _This is not a block for getting this PR merged. We can discuss it separately._

File: .github/workflows/test-build-deploy.yml (hunk 8 above)
Comment (pracucci) [reply .github/workflows/test-build-deploy.yml:101]: I also think we don't need to store these artifacts anymore.

File: .github/workflows/test-build-deploy.yml (hunk 12 above)
Comment (pracucci) [reply .github/workflows/test-build-deploy.yml:21]: But `BUILD_IN_CONTAINER` is not an environment variable. At least, doesn't work if you set it as env variable. Am I missing anything?

**Review by pracucci — APPROVED** [review]
LGTM, thanks!

**Comment by pracucci** [comment]
@AzfaarQureshi We merged the PR since it's in a good shape, but we would be glad if you could address the remaining comments in a subsequent PR. Thanks!

**Comment by pstibrany** [comment]
> But BUILD_IN_CONTAINER is not an environment variable. At least, doesn't work if you set it as env variable. Am I missing anything?

"Every environment variable that make sees when it starts up is transformed into a make variable with the same name and value." (from [GNU Make docs](https://www.gnu.org/software/make/manual/html_node/Environment.html))

File: .github/README.md (hunk 10 above)
Comment (shovnik) [reply .github/README.md:82]: The compression was in fact unnecessary. We removed it in the subsequent migration PR: https://github.com/cortexproject/cortex/pull/3341

File: .github/workflows/test-build-deploy.yml (hunk 11 above)
Comment (shovnik) [reply .github/workflows/test-build-deploy.yml:16]: Added a comment explaining this within the code itself in the subsequent PR to address this.
https://github.com/cortexproject/cortex/pull/3341
